src/osd/PrimaryLogPG.cc

```diff
@@ -2969,6 +3042,171 @@ void PrimaryLogPG::do_proxy_write(OpRequestRef op, const hobject_t& missing_oid,
   in_progress_proxy_ops[soid].push_back(op);
 }
 
+void PrimaryLogPG::do_chunk_op(OpRequestRef op, const hobject_t& missing_oid, 
+			       ObjectContextRef obc, bool write_ordered)
+{
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  OSDOp *osd_op = NULL;
+  for (unsigned int i = 0; i < m->ops.size(); i++) {
+    osd_op = &m->ops[i];
+    uint64_t cursor = osd_op->op.extent.offset;
+    uint64_t op_length = osd_op->op.extent.offset + osd_op->op.extent.length;
+    uint64_t chunk_length = obc->obs.oi.manifest.chunk_map[0].length;
+    uint64_t padding;
+    assert(chunk_length);
+    while (cursor < op_length) {
+      padding = cursor % chunk_length;
+      padding = chunk_length - padding;
+      uint64_t next_length =  padding;
+      if (cursor + next_length > op_length) {
+	next_length = op_length - cursor;
+      }
+      if (op->may_write() || write_ordered) {
+	do_chunk_write(op, obc, i, cursor, next_length);
+      } else {
+	do_chunk_read(op, obc, i, cursor, next_length);
+      }
+      cursor += chunk_length;
+    }
+  } 
+}
+
+void PrimaryLogPG::do_chunk_write(OpRequestRef op, ObjectContextRef obc, int op_index,
+				  uint64_t req_offset, uint64_t req_length)
+{
+  // NOTE: non-const because ProxyWriteOp takes a mutable ref
+  MOSDOp *m = static_cast<MOSDOp*>(op->get_nonconst_req());
+  SnapContext snapc(m->get_snap_seq(), m->get_snaps());
+  object_manifest_t *manifest = &obc->obs.oi.manifest;
+  uint64_t chunk_length = manifest->chunk_map[0].length;
+  uint64_t chunk_index = req_offset - (req_offset % chunk_length);
+
+  if (!manifest->chunk_map.count(chunk_index)) {
+    hobject_t ori_soid = m->get_hobj();
+    string chunk_name = ori_soid.oid.name;
+    chunk_name = chunk_name + "_" + to_string(chunk_index);
+    manifest->chunk_map[chunk_index] = manifest->chunk_map[0];
+    manifest->chunk_map[chunk_index].oid.oid = object_t(chunk_name);
+    manifest->chunk_map[chunk_index].flags = chunk_info_t::CHUNK_CLEAN;
+  } else {
+    manifest->chunk_map[chunk_index].flags = chunk_info_t::CHUNK_CLEAN;
+  }
+
+  hobject_t soid = manifest->chunk_map[chunk_index].oid;
+  hobject_t ori_soid = m->get_hobj();
+  object_locator_t oloc(soid);
+
+  unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY;
+  if (!(op->may_write() || op->may_cache())) {
+    flags |= CEPH_OSD_FLAG_RWORDERED;
+  }
+  dout(10) << __func__ << " Start do chunk proxy write for " << *m 
+	   << " index: " << op_index << " oid: " << soid.oid.name << " req_offset: " << req_offset 
+	   << " req_length: " << req_length << dendl;
+
+  osd_reqid_t reqid = m->get_reqid();
+  reqid.inc+=req_offset;
```

Comment:
I don't think this will work.  inc == incarnation, and it's there because the mds entity_name_t is something like mds.0, mds.1, and it's re-used after failover (with inc changing).  We could change that so that the requests come from mds.$gid (which is unique) instead of mds.$rank), and then repurpose the inc field for this.. that is probably my vote, actually.  @jcsp does that work for you?